pkg/workceptor/kubernetes.go

```diff
@@ -263,182 +273,326 @@ func (kw *kubeUnit) createPod(env map[string]string) error {
-				}
-			}
-		}()
-	} else {
-		kw.UpdateBasicStatus(WorkStateRunning, "Pod Running", stdout.Size())
-	}
+	// goroutine to cancel stdout stream
+	go func() {
+		select {
+		case <-kw.ctx.Done():
+			stdout.writer.Close()
+
+			return
+		case <-stdinErrChan:
+			stdout.writer.Close()
+
+			return
+		case <-finishedChan:
+			return
+		}
+	}()
 
-	// Actually run the streams.  This blocks until the pod finishes.
-	var errStdin error
-	var errStdout error
 	streamWait := sync.WaitGroup{}
 	streamWait.Add(2)
+
 	if skipStdin {
+		kw.UpdateBasicStatus(WorkStateRunning, "Pod Running", stdout.Size())
 		streamWait.Done()
 	} else {
 		go func() {
+			defer streamWait.Done()
+
+			kw.UpdateFullStatus(func(status *StatusFileData) {
+				status.State = WorkStatePending
+				status.Detail = "Sending stdin to pod"
+			})
+
+			var err error
 			for retries := 5; retries > 0; retries-- {
-				errStdin = exec.Stream(remotecommand.StreamOptions{
+				err = exec.Stream(remotecommand.StreamOptions{
 					Stdin: stdin,
 					Tty:   false,
 				})
-				if errStdin != nil {
-					logger.Warning("Problem opening stdin to pod %s, unit %s. Retrying.", kw.pod.Name, kw.unitID)
-					time.Sleep(time.Second * 5)
+				if err != nil {
+					// NOTE: io.EOF for stdin is handled by remotecommand and will not trigger this
+					kw.Warning("Error streaming stdin to pod %s/%s. Retrying: %s",
+						podNamespace,
+						podName,
+						err,
+					)
+					time.Sleep(100 * time.Millisecond)
 				} else {
 					break
 				}
 			}
-			if errStdin != nil {
-				logStream.Close()
+
+			if err != nil {
+				stdinErr = err
+				errMsg := fmt.Sprintf("Error streaming stdin to pod %s/%s: %s",
+					podNamespace,
+					podName,
+					err,
+				)
+				kw.Error(errMsg)
+				kw.UpdateBasicStatus(WorkStateFailed, errMsg, stdout.Size())
+
+				close(stdinErrChan) // signal STDOUT goroutine to stop
+			} else {
+				if stdin.Error() == io.EOF {
+					kw.UpdateBasicStatus(WorkStateRunning, "Pod Running", stdout.Size())
+				} else {
+					// this is probably not possible...
+					errMsg := fmt.Sprintf("Error reading stdin: %s", stdin.Error())
+					kw.Error(errMsg)
+					kw.UpdateBasicStatus(WorkStateFailed, errMsg, stdout.Size())
+
+					close(stdinErrChan) // signal STDOUT goroutine to stop
+				}
 			}
-			streamWait.Done()
 		}()
 	}
+
+	// TODO: @seth continue and go through this goroutine
 	go func() {
-		_, errStdout = io.Copy(stdout, logStream)
-		streamWait.Done()
+		defer streamWait.Done()
+
+		var sinceTime time.Time
+		var logStream io.ReadCloser
+		numEOF := 0 // resets to 0 on each successful read from pod stdout
+
+		for {
+			if stdinErr != nil {
+				break
+			}
+
+			// get pod, with retry
+			for retries := 5; retries > 0; retries-- {
+				kw.pod, err = kw.clientset.CoreV1().Pods(podNamespace).Get(kw.ctx, podName, metav1.GetOptions{})
+				if err == nil {
+					break
+				} else {
+					time.Sleep(100 * time.Millisecond)
```

Comment:
On my machine, this wasn't quite long enough to recover gracefully when the kubelet is restarted. Changing this to 1 second made it work. I see where we changed 4 other places from 5 seconds to 100 milliseconds. Let's go with somewhere in the middle... maybe 1 or 2 seconds.

Reply:
logger.warn here too.